File: solura-eco/backend/app/auth/session.py
"""Signed session tokens -- format matches Cortege's session.ts / Tender
Agent's session.py: base64url(payload).hex(hmac). Ported here, not
reinvented, per the multi-tenant SaaS playbook (see the Solura wiki).

The HMAC is computed over the raw base64url-decoded payload bytes, not a
re-serialized JSON string -- this is what lets the Next.js frontend verify
the same token without needing byte-identical JSON serialization between
Python and JS (it never re-encodes the payload, only decodes+verifies it).
"""
import base64
import hashlib
import hmac
import json
import time
from typing import Optional

DEFAULT_MAX_AGE_SECONDS = 30 * 24 * 3600  # 30 days


def _b64url_encode(data: bytes) -> str:
    return base64.urlsafe_b64encode(data).rstrip(b"=").decode("ascii")


def _b64url_decode(s: str) -> bytes:
    padding = "=" * (-len(s) % 4)
    return base64.urlsafe_b64decode(s + padding)


def create_session_token(member_id: str, username: str, secret: str) -> str:
    payload = {
        "member_id": member_id,
        "username": username,
        "issued_at": int(time.time()),
    }
    payload_bytes = json.dumps(payload, separators=(",", ":")).encode("utf-8")
    payload_b64 = _b64url_encode(payload_bytes)
    sig = hmac.new(secret.encode("utf-8"), payload_bytes, hashlib.sha256).hexdigest()
    return f"{payload_b64}.{sig}"


def verify_session_token(
    token: Optional[str], secret: str, max_age_seconds: int = DEFAULT_MAX_AGE_SECONDS
) -> Optional[dict]:
    if not token or "." not in token:
        return None

    payload_b64, sig = token.split(".", 1)

    try:
        payload_bytes = _b64url_decode(payload_b64)
    except Exception:
        return None

    expected_sig = hmac.new(secret.encode("utf-8"), payload_bytes, hashlib.sha256).hexdigest()
    if not hmac.compare_digest(sig, expected_sig):
        return None

    try:
        payload = json.loads(payload_bytes)
    except Exception:
        return None

    if time.time() - payload.get("issued_at", 0) > max_age_seconds:
        return None

    return payload

File: solura-eco/frontend/src/lib/session.ts
// solura-eco/frontend/src/lib/session.ts
//
// Verifies the same base64url(payload).hex(hmac) token format the backend
// issues (app/auth/session.py) -- using Web Crypto (crypto.subtle) so this
// runs in both the Edge runtime (proxy.ts) and Node (the /api/login route
// handler) without a native crypto dependency. Never re-serializes the
// payload to JSON before hashing -- it hashes the raw decoded bytes, which
// is what lets this match Python's HMAC without needing byte-identical
// JSON formatting between the two languages.

const MAX_AGE_SECONDS = 30 * 24 * 3600; // 30 days, matches the backend

export type SessionPayload = {
  member_id: string;
  username: string;
  issued_at: number;
};

function b64urlDecode(s: string): Uint8Array {
  const padded = s + "=".repeat((4 - (s.length % 4)) % 4);
  const base64 = padded.replace(/-/g, "+").replace(/_/g, "/");
  const binary = atob(base64);
  const bytes = new Uint8Array(binary.length);
  for (let i = 0; i < binary.length; i++) bytes[i] = binary.charCodeAt(i);
  return bytes;
}

function toHex(buffer: ArrayBuffer): string {
  return Array.from(new Uint8Array(buffer))
    .map((b) => b.toString(16).padStart(2, "0"))
    .join("");
}

export async function verifySessionToken(
  token: string | undefined,
  secret: string
): Promise<SessionPayload | null> {
  if (!token || !token.includes(".")) return null;

  const [payloadB64, sig] = token.split(".");
  let payloadBytes: Uint8Array;
  try {
    payloadBytes = b64urlDecode(payloadB64);
  } catch {
    return null;
  }

  const key = await crypto.subtle.importKey(
    "raw",
    new TextEncoder().encode(secret),
    { name: "HMAC", hash: "SHA-256" },
    false,
    ["sign"]
  );
  const expectedSigBuf = await crypto.subtle.sign(
    "HMAC",
    key,
    payloadBytes.slice().buffer
  );
  const expectedSig = toHex(expectedSigBuf);

  if (expectedSig.length !== sig.length) return null;
  let diff = 0;
  for (let i = 0; i < expectedSig.length; i++) {
    diff |= expectedSig.charCodeAt(i) ^ sig.charCodeAt(i);
  }
  if (diff !== 0) return null;

  let payload: SessionPayload;
  try {
    payload = JSON.parse(new TextDecoder().decode(payloadBytes)) as SessionPayload;
  } catch {
    return null;
  }

  if (Date.now() / 1000 - payload.issued_at > MAX_AGE_SECONDS) return null;

  return payload;
}
